### DISCUSSION TITLE
Packing optimised for volume?

### DISCUSSION THREAD

**Discussion by billj9000** [dvdoug/BoxPacker#600] [Q&A] [answered]
I've just started exploring BoxPacker but it looks like it will be very useful for our ecommerce application.
I've been tinkering with the first example in the documentation and have added a box between the smallest and largest in size.

    /*
     * Add choices of box type - in this example the dimensions are passed in directly via constructor,
     * but for real code you would probably pass in objects retrieved from a database instead
     */
    $packer->addBox(
        new TestBox(
            reference: 'Le petite box',
            outerWidth: 300,
            outerLength: 300,
            outerDepth: 10,
            emptyWeight: 10,
            innerWidth: 296,
            innerLength: 296,
            innerDepth: 8,
            maxWeight: 1000
        )
    );
    $packer->addBox(
        new TestBox(
            reference: 'Le medium box',
            outerWidth: 300,
            outerLength: 300,
            outerDepth: 50,
            emptyWeight: 10,
            innerWidth: 296,
            innerLength: 296,
            innerDepth: 48,
            maxWeight: 1000
        )
    );
    $packer->addBox(
        new TestBox(
            reference: 'Le grande box',
            outerWidth: 3000,
            outerLength: 3000,
            outerDepth: 100,
            emptyWeight: 100,
            innerWidth: 2960,
            innerLength: 2960,
            innerDepth: 80,
            maxWeight: 10000
        )
    );

I haven't changed the items in the example, just added the new box.
When I run this, it packs everything into 1 x Le grande box, the 3m x 3m x 10cm box. However, if I remove that box from the list, it packs the items into 2 x Le medium box. Since two medium boxes occupy 1% of the volume of the large box, I obviously would prefer the two medium boxes. The docs say that BoxPacker treats volume as a proxy for cost and I think it says somewhere that it optimises for cost so how can I make it prefer to pack 2 x medium boxes instead of 1 x huge box?
Do I have to pack all permutations and then optimise the results myself?

**Comment by billj9000** [comment]
Looking through the documentation again, I see:

> By default, BoxPacker will optimise for the largest number of items in a box, with volume acting as a tie-breaker.

So it looks as though the behaviour I see above is indeed the default and I need to create a custom PackedBoxSorter.

**Reply by dvdoug** [reply, marked as answer]
Yes, correct - in my experience it's usually cheaper to use 1 oversized box with lots of wasted space rather than multiple smaller boxes so that's what the behaviour defaults to.

But everything is configurable as you've found 👍

**Reply by billj9000** [reply]
Thanks for the reply, Doug. It depends on the application I think and the economics of international freight are different from domestic freight where a flat rate often applies. For our business, if a customer orders a long, narrow item, say 1.1m long, we can send that in a 1.2m x 100mm postal tube to the US for about £28. If he orders a different item of around 150mm x 150mm x 120mm then that also won't cost very much to ship to the US. However, if he orders both together, the smallest standard box we have that will accommodate both parts costs around £480 to ship to the US. In that case, it is much cheaper to send two packages but our current homebrew box-packing system won't handle multiple packages.

I'll look at customising BoxPacker to try to optimise things as we need them. It looks as though it can be customised to optimise the choice of an individual box but I'm not sure how to optimise it for the permutation. Perhaps it is better simply to get all the permutations and select the optimal one outside BoxPacker by comparing actual shipping costs, which we can do automatically. Very few of our orders contain more than 10 items and most of them a lot fewer so perhaps the number of permutations would be manageable.

**Reply by dvdoug** [reply]
Ah, tubes do break all assumptions 🤣

£480 is a lot though, that's a huge pricing gap indeed. I think your described approach is the one that I'd go with myself, much cheaper to to spend the CPU time on some extra permutations than waste that many £££ on shipping

**Comment by billj9000** [comment]
> Ah, tubes do break all assumptions 🤣

We treat a tube as a long rectangular cuboid and this approximation works well enough most of the time. So in this case a tube is just a box.
